### PR TITLE
Support deletion of documents when a connector is deleted

### PR DESCRIPTION
**Pull request by Weves**

### CODE REVIEW

File: backend/danswer/db/models.py [hunk 1]
```diff
@@ -190,3 +213,94 @@ def __repr__(self) -> str:
...
+
+    connector: Mapped[Connector] = relationship(
+        "Connector", back_populates="documents_by_connector"
+    )
+    credential: Mapped[Credential] = relationship(
+        "Credential", back_populates="documents_by_credential"
+    )
+
+
+class DocumentStoreEntry(Base):
```
Comment (Weves) [inline backend/danswer/db/models.py]: rename to chunk

File: backend/danswer/background/connector_deletion.py [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+        num_docs_deleted = 0
+        document_store_entries_to_delete = (
+            get_document_store_entries_with_single_connector_credential_pair(
+                db_session=db_session,
+                connector_id=connector_id,
+                credential_id=credential_id,
+            )
+        )
+        if document_store_entries_to_delete:
+            vector_index.delete(
```
Comment (Weves) [inline backend/danswer/background/connector_deletion.py]: fix IDs

File: backend/danswer/background/connector_deletion.py [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+            # pair that's still around with the same permissions), skip
+            if old_allowed_users == allowed_users:
+                continue
+
+            update_groups[allowed_users] = update_groups.get(allowed_users, []) + [
+                document_id
+            ]
+
+        # actually perform the updates in the document store
+        update_requests = [
```
Comment (Weves) [inline backend/danswer/background/connector_deletion.py:145]: look at this

File: backend/danswer/datastores/qdrant/store.py [hunk 4]
```diff
@@ -78,10 +86,14 @@ def __init__(self, collection: str = QDRANT_DEFAULT_COLLECTION) -> None:
         self.collection = collection
         self.client = get_qdrant_client()
 
-    def index(self, chunks: list[EmbeddedIndexChunk], user_id: UUID | None) -> int:
+    def index(
```
Comment (Weves) [inline backend/danswer/datastores/qdrant/store.py:89]: make sure that this works for mini-chunking

File: backend/danswer/background/connector_deletion.py [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+
+    def _cleanup() -> None:
+        # cleanup everything else up
+        # we cannot undue the deletion of the document store entries if something
+        # goes wrong since they happen outside the postgres world. Best we can do
+        # is keep everything else around and mark the deletion attempt as failed.
+        # If it's a transient failure, re-deleting the connector / credential pair should
+        # fix the weird state.
+        # TODO: lock anything to do with this connector via transaction isolation
+        delete_index_attempts(
```
Comment (Weves) [inline backend/danswer/background/connector_deletion.py:177]: talk about foreign key constraint

File: backend/danswer/server/manage.py [hunk 6]
```diff
@@ -375,6 +416,48 @@ def delete_openai_api_key(
     get_dynamic_config_store().delete(OPENAI_API_KEY_STORAGE_KEY)
 
 
+@router.post("/admin/deletion-attempt")
+def create_deletion_attempt_for_connector_id(
```
Comment (Weves) [inline backend/danswer/server/manage.py:411]: Check for disabled

File: web/src/components/admin/connectors/table/DeleteColumn.tsx [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+            type: "success",
+          });
+        }
+        setTimeout(() => {
+          setPopup(null);
+        }, 4000);
+        onUpdate();
+      }}
+    >
+      <TrashIcon />
```
Comment (Weves) [inline web/src/components/admin/connectors/table/DeleteColumn.tsx]: Add tooltip

File: backend/danswer/background/connector_deletion.py [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+        # update it's access rather than outright delete it
+        document_by_connector_credential_pairs_to_update = (
+            get_document_by_connector_credential_pairs_indexed_by_multiple(
+                db_session=db_session,
+                connector_id=connector_id,
+                credential_id=credential_id,
+            )
+        )
+
+        def _get_user(
```
Comment (yuhongsun96) [inline backend/danswer/background/connector_deletion.py:103]: slight preference to have these helper functions declared at the top of the function rather than interwoven into it

File: backend/danswer/background/connector_deletion.py [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+        ) in document_by_connector_credential_pairs_to_update:
+            document_id = document_by_connector_credential_pair.id
+            user = _get_user(document_by_connector_credential_pair.credential)
+            document_id_to_allowed_users[document_id].append(user)
+
+            # if there's another connector / credential pair which has indexed this
+            # document with the same access, we don't need to update it since removing
+            # the access from this connector / credential pair won't change anything
+            if (
+                document_by_connector_credential_pair.connector_id != connector_id
```
Comment (yuhongsun96) [inline backend/danswer/background/connector_deletion.py:127]: this 4 layer function nesting with inherited scopes is really hard to follow at this point but ok... I understand how it works I guess

File: backend/danswer/background/connector_deletion.py [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+        document_id_to_allowed_users: dict[str, list[str]] = defaultdict(list)
+        for (
+            document_by_connector_credential_pair
+        ) in document_by_connector_credential_pairs_to_update:
+            document_id = document_by_connector_credential_pair.id
+            user = _get_user(document_by_connector_credential_pair.credential)
+            document_id_to_allowed_users[document_id].append(user)
+
+            # if there's another connector / credential pair which has indexed this
+            # document with the same access, we don't need to update it since removing
```
Comment (yuhongsun96) [inline backend/danswer/background/connector_deletion.py:124]: clearer to say with credential belonging to the same user as the credential being deleted, but whatever eh? Also bit iffy if a credential belongs to multiple users at some future point, but ya nvm its good for now

File: backend/danswer/background/connector_deletion.py [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+        delete_document_by_connector_credential_pair(
+            db_session=db_session,
+            document_ids=list(document_id_to_allowed_users.keys()),
+        )
+
+    _update_multi_indexed_docs()
+
+    def _cleanup() -> None:
+        # cleanup everything else up
+        # we cannot undue the deletion of the document store entries if something
```
Comment (yuhongsun96) [inline backend/danswer/background/connector_deletion.py]: undue? :|

File: backend/danswer/connectors/utils.py [hunk 12]
```diff
@@ -1,19 +1,20 @@
...
+from collections.abc import Iterable
 from itertools import islice
 from typing import TypeVar
 
 T = TypeVar("T")
 
 
 def batch_generator(
-    generator: Iterator[T],
+    items: Iterable[T],
```
Comment (yuhongsun96) [inline backend/danswer/connectors/utils.py:11]: is this a more standard naming

**Review by yuhongsun96 — APPROVED** [review]
